pipeline/lib/metric.ml

```diff
@@ -0,0 +1,230 @@
+
+let calc_diff v cv =
+  if cv <> 0.
+  then (v -. cv) /. cv *. 100.
+  else if v <> 0.
+  then (v -. cv) /. v *. 100.
+  else 0.
+
+let find_changed_metrics ~metrics ~compare_metrics =
+  BenchmarksData.merge
+    (fun _ (value : Current_bench_json.Latest.value option)
+         (compare_value : Current_bench_json.Latest.value option) ->
+      match (value, compare_value) with
+      | Some value, Some compare_value ->
+          let delta =
+            calc_diff (avg_of_value value) (avg_of_value compare_value)
+          in
+          if Float.abs delta > change_threshold then Some delta else None
+      | Some _, None -> Some Float.nan
+      | _ -> None)
+    metrics compare_metrics
+
+module SeenBenchmarks = Set.Make (MetricKey)
+
+let format_changes ~changes =
+  let text_list, _ =
+    BenchmarksData.fold
+      (fun (bench_name, test_name, metric_name) change (acc, seen) ->
+        let bench_key = (bench_name, "", "") in
+        let test_key = (bench_name, test_name, "") in
+        let bench_seen = SeenBenchmarks.mem bench_key seen in
+        let test_seen = SeenBenchmarks.mem test_key seen in
+        let entry, seen =
+          match bench_seen with
+          | false ->
+              ( Fmt.str "\n\n## Benchmark: %s\n" bench_name,
+                SeenBenchmarks.add bench_key seen )
+          | true -> ("", seen)
+        in
+        let entry, seen =
+          match test_seen with
+          | false ->
+              ( Fmt.str "%s\n### Test: %s\n" entry test_name,
+                SeenBenchmarks.add test_key seen )
+          | true -> (entry, seen)
+        in
+        let change =
+          if Float.is_nan change
+          then Fmt.str "- %s is a new metric" metric_name
+          else Fmt.str "- %s changed by %f%%" metric_name change
+        in
+        let entry = Fmt.str "%s%s" entry change in
+        (entry :: acc, seen))
+      changes ([], SeenBenchmarks.empty)
+  in
+  String.concat "\n" (List.rev text_list)
+
+let commit_url ~repo_id ~commit =
+  let url = Fmt.str "https://github.com/%s/commit/%s" repo_id commit in
+  let short_sha = String.sub commit 0 7 in
+  Fmt.str "[%s](%s)" short_sha url
+
+let pull_url ~repo_id ~pull_number =
+  let url = Fmt.str "https://github.com/%s/pulls/%d" repo_id pull_number in
+  let pull = Fmt.str "#%d" pull_number in
+  Fmt.str "[%s](%s)" pull url
+
+let format_changes_message ~changes ~repository ~worker ~docker_image
+    ~main_commit =
+  let pull_number = Repository.pull_number repository |> Option.get in
+  let commit = Repository.commit_hash repository in
+  let repo_id = Repository.info repository in
+  let main_commit = Option.value ~default:"Unknown hash" main_commit in
+  let header =
+    Fmt.str
+      "%s (%s) changes the metrics as follows in comparison to `main` (%s) \
+       when running on `%s (%s)`:"
+      (pull_url ~repo_id ~pull_number)
+      (commit_url ~repo_id ~commit)
+      (commit_url ~repo_id ~commit:main_commit)
+      worker docker_image
```

Comment:
Can we add a quick link to the benchmarks webpage for this worker/image? (otherwise it's a bit painful to access it ^^')